[Flaking Test] UT k8s.io/component-base/metrics: legacyregistry

### Which jobs are flaking?

https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-ppc64le/1941058055126913024
https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-1-32/1937945831206293504
https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-dependencies/1937216429380079616

### Which tests are flaking?

`/k8s.io/component-base/metrics/legacyregistry/registry_test.go:TestProcessStartTimeHeader`

### Since when has it been flaking?

Triage tool shows the earliest failures on [6/23/2025, 11:59:53 PM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-dependencies/1937216429380079616) as per [Link](https://storage.googleapis.com/k8s-triage/index.html?date=2025-07-04&test=k8s.io%2Fcomponent-base%2Fmetrics%2Flegacyregistry). 

### Testgrid link

_No response_

### Reason for failure (if possible)

   
```
k8s.io/component-base/metrics: legacyregistry 
{Failed  === RUN   TestProcessStartTimeHeader     registry_test.go:41: got 1751621652, wanted 1751621653 --- FAIL: TestProcessStartTimeHeader (0.02s) }

```

### Anything else we need to know?

We were able to reproduce this flake locally using stress tool.
**_Steps to reproduce:_** 
1. Install stress tool (# go install golang.org/x/tools/cmd/stress@latest).
2. Clone the k8s repo (git clone https://github.com/kubernetes/kubernetes.git).
3. cd kubernetes
4.  go test ./staging/src/k8s.io/component-base/metrics/legacyregistry/ -c race
5.  /root/go/bin/stress ./legacyregistry.test -test.run TestProcessStartTimeHeader

```
[root@varsharani1 kubernetes]# uname -a
Linux varsharani1.fyre.ibm.com 5.14.0-585.el9.x86_64 #1 SMP PREEMPT_DYNAMIC Wed May 14 18:37:27 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux
[root@varsharani1 kubernetes]#

```
```
[root@varsharani1 kubernetes]# /root/go/bin/stress ./legacyregistry.test -test.run TestProcessStartTimeHeader
5s: 3130 runs so far, 0 failures, 8 active
10s: 6282 runs so far, 0 failures, 8 active
15s: 9578 runs so far, 0 failures, 8 active

/tmp/go-stress-20250711T030443-2999400201
--- FAIL: TestProcessStartTimeHeader (0.00s)
    registry_test.go:41: got 1752228298, wanted 1752228299
FAIL


ERROR: exit status 1


/tmp/go-stress-20250711T030443-2982120025
--- FAIL: TestProcessStartTimeHeader (0.00s)
    registry_test.go:41: got 1752228299, wanted 1752228300
FAIL


ERROR: exit status 1

20s: 12780 runs so far, 2 failures (0.02%), 8 active
25s: 15825 runs so far, 2 failures (0.01%), 8 active

```

### Relevant SIG(s)

/sig testing

Able to reproduce it on `arm64` arch as well with stress tool,

```
%stress ./legacyregistry.test -test.run ^TestProcessStartTimeHeader
5s: 2849 runs so far, 0 failures, 10 active


10s: 6140 runs so far, 0 failures, 10 active
15s: 9322 runs so far, 0 failures, 10 active
20s: 12587 runs so far, 0 failures, 10 active

/var/folders/3p/5y2c1zc51x173nqxbdslj7200000gn/T/go-stress-20250801T115621-573344968
--- FAIL: TestProcessStartTimeHeader (0.00s)
    registry_test.go:41: got 1754029604, wanted 1754029605
FAIL


ERROR: exit status 1

```

It could be because of minor execution delay between handler initialisation and `now` under stress conditions. I will submit a PR to fix it.

--- a/staging/src/k8s.io/component-base/metrics/legacyregistry/registry.go
+++ b/staging/src/k8s.io/component-base/metrics/legacyregistry/registry.go
@@ -90,3 +90,8 @@ func CustomMustRegister(cs ...metrics.StableCollector) {
 		prometheus.MustRegister(c)
 	}
 }
+
+// GetProcessStart return processStart value
+func GetProcessStart() time.Time {
+	return processStart
+}

--- a/staging/src/k8s.io/component-base/metrics/legacyregistry/registry_test.go
+++ b/staging/src/k8s.io/component-base/metrics/legacyregistry/registry_test.go
@@ -21,23 +21,34 @@ import (
 	"net/http/httptest"
 	"strconv"
 	"testing"
-	"time"
 )
 
 const (
 	processStartTimeHeader = "Process-Start-Time-Unix"
 )
 
+var (
+	processStartTestCopy = &processStart
+)
+
 func TestProcessStartTimeHeader(t *testing.T) {
-	now := time.Now()
+	now := GetProcessStart()
 	handler := Handler()
 
-	request, _ := http.NewRequest("GET", "/", nil)
+	request, _ := http.NewRequest(http.MethodGet, "/", nil)
 	writer := httptest.NewRecorder()
 	handler.ServeHTTP(writer, request)
 	got := writer.Header().Get(processStartTimeHeader)
 	gotInt, _ := strconv.ParseInt(got, 10, 64)
 	if gotInt != now.Unix() {
 		t.Errorf("got %d, wanted %d", gotInt, now.Unix())
+	}
+}
+
+// processStart must never be reassigned after init.
+// This test ensures processStart doesn't change across calls
+func TestProcessStartImmutable(t *testing.T) {
+	if *processStartTestCopy != processStart {
+		t.Errorf("processStart test values differ: %v != %v", processStartTestCopy, processStart)
 	}
 }

